### PR TITLE
Implement a delta-debugging bitmap & better DAG shrinking

### PR DESCRIPTION
**Pull request by matheus23**
This implements two major things:
1. A proptest `ValueTree` and `Strategy` implementation for `RoaringBitmap` which shrinks using the [delta debugging algorithm](https://www.st.cs.uni-saarland.de/papers/tse2002/tse2002.pdf). This algorithm shrinks faster than the standard proptest algorithm in the best case (a single bit is causing a test failure), by throwing away multiple bits at a time (it looks like a binary search). But it also finds the minimal test case in the general case and has similar performance in that case.
2. A new proptest `ValueTree` and `Strategy` implementation for `arb_dag`, which first tries to shrink the vertices in the DAG (trying to remove more and more vertices that aren't contributing to a test failure) and then proceeds trying to reduce edges. This often makes it find the minimal DAG causing a test failure.

Let me know what you think!

Some things I wasn't quite sure about:
- I use rustfmt, but I tried to not make it format any existing code. LMK if I missed some place
- I made `delta_debugging_bitmap` public in the API, similar to how `strictly_upper_triangular_logical_matrix` is public, but I can see that being a stretch. LMK if you'd prefer it private (or perhaps even in a separate package?).
- The `arb_dag` API is now changed such that you can provide the vertex count range that you prefer, similar to e.g. the `vec` API in proptest. Previously it was also weighted towards generating the empty DAG, should I add something similar? Something that makes it prefer the edges of the interval perhaps?
- Currently the `*ValueTree` structs and their `impl`s are `pub`, similar to how they're pub in `proptest`. Would you prefer them to be `pub(crate)` instead?

Also some suggestions:
- Do you want to enable rustfmt?
- I think it's possible to change the return type for some functions from `Box<dyn Iterator<...>>` to `impl Iterator<...> + '_`
- Some places use `vec![...].iter().cloned()`, but you can skip a clone by going `vec![...].into_iter()` directly.

Happy to work on a PR with any of these suggestions you agree with :)

### CODE REVIEW

**Comment by adaszko** [comment]
This is rad!  Give me a day or two to muster some free time to understand this properly please

File: src/dag.rs [hunk 1]
```diff
@@ -388,28 +394,119 @@ impl DirectedAcyclicGraph {
...
+    vertex_count: Range<Vertex>,
+}
+
+#[derive(Debug)]
+pub struct DirectedAcyclicGraphValueTree {
+    vertex_count: u16,
+    vertex_mask_tree: DeltaDebuggingBitmapValueTree,
+    edge_bitmap_tree: DeltaDebuggingBitmapValueTree,
+    start_simplifying_edges: bool,
+}
```
Comment (adaszko) [inline src/dag.rs:452]: Suggestion: To improve readability, could this struct be made into an enum so that it's more apparent it's a state machine?  Something like (as an inner enum if need be):

```Rust
enum SimplifyingState {
    SimplifyingVertices(vertex_mask_tree, edge_bitmap_tree), // not sure if both are necessary
    SimplifyingEdges(vertex_mask_tree, edge_bitmap_tree), // not sure if both are necessary
}
```

File: src/dag.rs [hunk 2]
```diff
@@ -388,28 +394,119 @@ impl DirectedAcyclicGraph {
...
+                self.vertex_count.start,
+                self.vertex_count.end,
+                self.vertex_count.start,
+                self.vertex_count.end
+            );
+        }
+        let vertex_count =
+            Uniform::new(self.vertex_count.start, self.vertex_count.end - 1).sample(runner.rng());
+        let bitmap_size = strictly_upper_triangular_matrix_capacity(vertex_count);
+        let iter = (0..bitmap_size as u32).filter(|_| runner.rng().gen_bool(0.5));
```
Comment (adaszko) [inline src/dag.rs]: Because of `gen_bool(0.5)`, I'd suggest renaming `DirectedAcyclicGraphStrategy` to `UniformlyDistributedEdgesStrategy` as this seems to be its defining trait.

File: src/dag.rs [hunk 3]
```diff
@@ -752,4 +854,54 @@ mod tests {
...
+        edge_bitmap.insert_range(0..bitmap_size);
+        let mut vertex_mask = RoaringBitmap::new();
+        vertex_mask.insert_range(0..vertex_count as u32);
+
+        let full_graph_tree = DirectedAcyclicGraphValueTree {
+            vertex_count,
+            edge_bitmap_tree: DeltaDebuggingBitmapValueTree::new(edge_bitmap),
+            vertex_mask_tree: DeltaDebuggingBitmapValueTree::new(vertex_mask),
+            start_simplifying_edges: false,
+        };
```
Comment (adaszko) [inline src/dag.rs]: It'd be nice to put this piece into a dedicated fn, something along the lines of:

```Rust
let dag = DirectedAcyclicGraph::fully_connected(10);
let value_tree = dag.as_adjacency_matrix();
```

That would make a semantic connection between value tree and DAG representation, while also providing a new useful pub fn on the `DirectedAcyclicGraph` type.

File: src/delta_debugging_bitmap.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    DeltaDebuggingBitmapStrategy {
+        size: size.into(),
+        ones_probability: 0.5,
+    }
+}
+
+pub fn arb_bitmap(
+    size: impl Into<Range<u32>>,
+    ones_probability: f64,
+) -> DeltaDebuggingBitmapStrategy {
```
Comment (adaszko) [inline src/delta_debugging_bitmap.rs]: To better fit mathematical definitions, it'd be better to name this fn `arb_bitmap_uniform()` :)  Note that a distribution with `ones_probability` of 0.3 is also uniform

File: src/delta_debugging_bitmap.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+            self.split_into = max(self.split_into - 1, 2);
+            self.complement = false;
+        } else if self.split_into == 1 {
+            // We add a special case to test the empty set as early as possible
+            self.complement = true;
+        } else {
+            // Successfully reduced to subset
+            self.split_into = 2;
+            self.complement = false;
+        }
```
Comment (adaszko) [inline src/delta_debugging_bitmap.rs:155]: I'd love if you could make this into a enum state machine too.  It really makes the code easier to change later

File: src/strictly_upper_triangular_logical_matrix.rs [hunk 6]
```diff
@@ -49,6 +52,22 @@ pub struct StrictlyUpperTriangularLogicalMatrix {
...
+    fn eq(&self, other: &Self) -> bool {
+        if self.size != other.size {
+            return false;
+        }
+        for (r, c) in RowColumnIterator::new(self.size) {
+            if self.get(r, c) != other.get(r, c) {
+                return false;
+            }
+        }
+        true
```
Comment (adaszko) [inline src/strictly_upper_triangular_logical_matrix.rs]: This should avoid all the multiplications and additions done for indexing sake:

```suggestion
        self.iter_ones().eq(other.iter_ones())
```

File: src/dag.rs [hunk 7]
```diff
@@ -388,28 +394,119 @@ impl DirectedAcyclicGraph {
...
+        ) as usize);
+        for from_src in 0..self.vertex_count {
+            if vertex_mask.contains(from_src as u32) {
+                let mut to_dst = from_dst + 1;
+                for to_src in from_src + 1..self.vertex_count {
+                    if vertex_mask.contains(to_src as u32) {
+                        let n = self.vertex_count as u32;
+                        let i = from_src as u32;
+                        let j = to_src as u32;
+                        let edge_idx = i * n + j - (((i + 1) * (i + 1) - (i + 1)) / 2 + (i + 1));
```
Comment (adaszko) [inline src/dag.rs]: This made me realize that during the dograph -> roaring-graphs transition, `StrictlyUpperTriangularLogicalMatrix` name lost its meaning since we're no longer using the "upper triangular" packing scheme.  I should rename it to `SparseSquareLogicalMatrix` since roaring bitmaps take care of storing only the relevant bits (literally bits).  I'm going to rename it in some future crate release.

In any case, this formula `i * n + j - (((i + 1) * (i + 1) - (i + 1)) / 2 + (i + 1));` made me pause for a bit.  Can we put it into a separate fn with a descriptive name, e.g. `strictly_upper_triangular_index` or similar?  🙏

**Review by adaszko — COMMENTED** [review]
> Let me know what you think!

As I said, this is absolutely rad!  I can see how it reduces tests execution time on failure if your DAGs are in the database or on disk.

> Some things I wasn't quite sure about:
> - I use rustfmt, but I tried to not make it format any existing code. LMK if I missed some place
> - Do you want to enable rustfmt?

It's all good.  Now that I'm not the only contributor, rustfmt is definitely welcome but let's leave it for a separate PR.  Otherwise it's hard to see what has changed in a PR.

> - I made `delta_debugging_bitmap` public in the API, similar to how `strictly_upper_triangular_logical_matrix` is public, but I can see that being a stretch. LMK if you'd prefer it private (or perhaps even in a separate package?).

I'm leaning towards keeping it private.  Have you considered contributing this shrinking strategy upstream to proptest?  If I'm reading [the code](https://github.com/proptest-rs/proptest/blob/658d5130e72bd14290d7c15f7affb38cda66522e/proptest/src/bits.rs#L314-L332) correctly, it does a linear scan over the bitmaps, so it could be improved with your technique?

> - The `arb_dag` API is now changed such that you can provide the vertex count range that you prefer, similar to e.g. the `vec` API in proptest. Previously it was also weighted towards generating the empty DAG, should I add something similar? Something that makes it prefer the edges of the interval perhaps?

It's great as it is 👍 

> - Currently the `*ValueTree` structs and their `impl`s are `pub`, similar to how they're pub in `proptest`. Would you prefer them to be `pub(crate)` instead?

Yes, going from `pub(crate)` to `pub` is backward-compatible change.  The other way around, is a breaking change.

> - I think it's possible to change the return type for some functions from `Box<dyn Iterator<...>>` to `impl Iterator<...> + '_`

I'm not sure about this one TBH, the `impl Trait` Rust feature isn't fully implemented in the compiler yet, so I'm conservative with its use.

> - Some places use `vec![...].iter().cloned()`, but you can skip a clone by going `vec![...].into_iter()` directly.

Oh, that's just sloppy coding on my part.  `.into_iter()` is definitely better.

> Happy to work on a PR with any of these suggestions you agree with :)

Appreciated! :)

File: src/dag.rs (hunk 2 above)
Comment (matheus23) [reply src/dag.rs]: Ah right. What do you think of making the probability of an edge configurable & keeping the DAG name?

Comment (matheus23) [reply src/dag.rs]: Nevermind. Still uniform :P

File: src/dag.rs (hunk 7 above)
Comment (matheus23) [reply src/dag.rs]: LOL yeah sorry for keeping that in, I meant to clean that up a bit :sweat_smile: 

Yeah, losing the packing scheme confused be a bunch, I initially wrote the code for dograph, so there was some head-scratching on my end.

File: src/dag.rs (hunk 3 above)
Comment (matheus23) [reply src/dag.rs]: Hm, the main issue with this is that the `edge_bitmap` is actually using the strictly upper packed matrix representation, whereas the DAG's `StrictlyUpperTriangularLogicalMatrix` doesn't use it anymore.

**Comment by matheus23** [comment]
There's actually a bunch of places where I wished I could've used the same packed representation that the original dograph library used (e.g. DirectedAcyclicGraphValueTree::current could use `DirectedAcyclicGraph::masked_subgraph` instead). What was the reasoning for abandoning that and instead going for the square representation?

File: src/delta_debugging_bitmap.rs (hunk 5 above)
Comment (matheus23) [reply src/delta_debugging_bitmap.rs:155]: Hmm. I'm generally a fan of representing state machines using `enum`s instead of booleans, but in this case it really complicates the implementation quite a bit. E.g. look at the `ValueTree::current` function.
With `Self` represented as:
```rs
pub enum DeltaDebuggingBitmapValueTree {
    CheckSubset {
        upper: RoaringBitmap,
        split_into: u64,
        split_index: u64,
    },
    CheckSubsetComplement {
        upper: RoaringBitmap,
        split_into: u64,
        split_index: u64,
    },
}
```

You end up needing to match as the very first thing in the function, to get access to `upper` and `split_into`.
The fact that it's checking the complement vs. not the complement is only needed in the innermost part of the function however.

I really think in this case it's better to represent it as is.

**Comment by adaszko** [comment]
> There's actually a bunch of places where I wished I could've used the same packed representation that the original dograph library used (e.g. DirectedAcyclicGraphValueTree::current could use `DirectedAcyclicGraph::masked_subgraph` instead). What was the reasoning for abandoning that and instead going for the square representation?

The main reason was that it's hard to compute (row, column) pair given just an index in the bitmap.  It requires inverting the formula `((2 * size - i - 1) * i) / 2 + j - i - 1`, in a loop most often.  It's just simpler computationally and conceptually to store it in a square matrix and delegate space optimization to RoaringBitmaps.  They're adaptive and do not store all the zeros in the bitmap, so the space usage should be roughly the same, yet they're more flexible than packed representation.

File: src/delta_debugging_bitmap.rs (hunk 5 above)
Comment (adaszko) [reply src/delta_debugging_bitmap.rs:155]: There might have been a slight misunderstanding :)

It's totally fine to lift the enum fields used by all states into the parent struct.  Only the state-specific stuff should go into enums.  Same with `ReductionState` :)

File: src/dag.rs (hunk 3 above)
Comment (adaszko) [reply src/dag.rs]: Right, nevermind then!

**Comment by adaszko** [comment]
> There's actually a bunch of places where I wished I could've used the same packed representation that the original dograph library used (e.g. DirectedAcyclicGraphValueTree::current could use `DirectedAcyclicGraph::masked_subgraph` instead). What was the reasoning for abandoning that and instead going for the square representation?

If it's going to make the code clearer, we can bring the packed representation data type back as a crate-private type, just to use it as an internal state in the ValueTree computations but produce roaring bitmaps as the final result.  Various graph algorithms can be more efficient when you can get back (row, colum) pair from an index, hence the square representation is the preferred one.

File: src/delta_debugging_bitmap.rs (hunk 5 above)
Comment (matheus23) [reply src/delta_debugging_bitmap.rs:155]: Ah okay - tackling some boolean blindness. I'm happy with doing that :)

Comment (adaszko) [reply src/delta_debugging_bitmap.rs:155]: > Ah okay - tackling some boolean blindness. I'm happy with doing that :)

Exactly, that's the name for it 👍

**Comment by matheus23** [comment]
Hm. Let me try switching it around: The ValueTree can possibly use the non-packed representation instead of the packed one.

**Comment by matheus23** [comment]
This should be better. Should be ready for a re-review @adaszko

**Comment by adaszko** [comment]
Great stuff 🎉  Thanks!
